File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/common/LiveDataImpl.kt
package dev.hebei.camerax_android.common

import androidx.camera.core.CameraState
import androidx.camera.core.ZoomState
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import dev.hebei.camerax_android.CameraStateApi
import dev.hebei.camerax_android.CameraXImpl
import dev.hebei.camerax_android.PigeonApiCameraStateLiveDataApi
import dev.hebei.camerax_android.PigeonApiTorchStateLiveDataApi
import dev.hebei.camerax_android.PigeonApiZoomStateLiveDataApi
import dev.hebei.camerax_android.TorchStateApi
import dev.hebei.camerax_android.core.api
import dev.hebei.camerax_android.core.torchStateApi

class CameraStateLiveDataImpl(private val impl: CameraXImpl) : PigeonApiCameraStateLiveDataApi(impl) {
    override fun getValue(pigeon_instance: CameraStateLiveData): CameraStateApi? {
        return pigeon_instance.value
    }

    override fun observe(pigeon_instance: CameraStateLiveData, observer: CameraStateObserver) {
        val owner = impl.activity as LifecycleOwner
        pigeon_instance.observe(owner, observer)
    }

    override fun removeObserver(pigeon_instance: CameraStateLiveData, observer: CameraStateObserver) {
        pigeon_instance.removeObserver(observer)
    }
}

class TorchStateLiveDataImpl(private val impl: CameraXImpl) : PigeonApiTorchStateLiveDataApi(impl) {
    override fun getValue(pigeon_instance: TorchStateLiveData): TorchStateApi? {
        return pigeon_instance.value
    }

    override fun observe(pigeon_instance: TorchStateLiveData, observer: TorchStateObserver) {
        val owner = impl.activity as LifecycleOwner
        return pigeon_instance.observe(owner, observer)
    }

    override fun removeObserver(pigeon_instance: TorchStateLiveData, observer: TorchStateObserver) {
        pigeon_instance.removeObserver(observer)
    }
}

class ZoomStateLiveDataImpl(private val impl: CameraXImpl) : PigeonApiZoomStateLiveDataApi(impl) {
    override fun getValue(pigeon_instance: ZoomStateLiveData): ZoomState? {
        return pigeon_instance.value
    }

    override fun observe(pigeon_instance: ZoomStateLiveData, observer: ZoomStateObserver) {
        val owner = impl.activity as LifecycleOwner
        pigeon_instance.observe(owner, observer)
    }

    override fun removeObserver(pigeon_instance: ZoomStateLiveData, observer: ZoomStateObserver) {
        pigeon_instance.removeObserver(observer)
    }
}

class CameraStateLiveData(internal val obj: LiveData<CameraState>) {
    val value get() = obj.value?.api

    fun observe(owner: LifecycleOwner, observer: CameraStateObserver) {
        obj.observe(owner, observer.obj)
    }

    fun removeObserver(observer: CameraStateObserver) {
        obj.removeObserver(observer.obj)
    }
}

class TorchStateLiveData(internal val obj: LiveData<Int>) {
    val value get() = obj.value?.torchStateApi

    fun observe(owner: LifecycleOwner, observer: TorchStateObserver) {
        obj.observe(owner, observer.obj)
    }

    fun removeObserver(observer: TorchStateObserver) {
        obj.removeObserver(observer.obj)
    }
}

class ZoomStateLiveData(internal val obj: LiveData<ZoomState>) {
    val value get() = obj.value

    fun observe(owner: LifecycleOwner, observer: ZoomStateObserver) {
        obj.observe(owner, observer.obj)
    }

    fun removeObserver(observer: ZoomStateObserver) {
        obj.removeObserver(observer.obj)
    }
}

File: camerax_ios/lib/src/common/live_data_impl.dart
import 'package:camerax_ios/src/api.dart';
import 'package:camerax_ios/src/common.dart';
import 'package:camerax_ios/src/core.dart';
import 'package:camerax_platform_interface/camerax_platform_interface.dart';

abstract base class LiveDataImpl<T> implements LiveData<T> {}

final class CameraStateLiveDataImpl extends LiveDataImpl<CameraState> {
  final CameraStateLiveDataProxyApi api;

  CameraStateLiveDataImpl.internal(this.api);

  @override
  Future<CameraState?> getValue() => api.getValue().then((e) => e?.impl);

  @override
  Future<void> observeForever(Observer<CameraState> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<CameraState> observer) =>
      api.removeObserver(observer.api);
}

final class TorchStateLiveDataImpl extends LiveDataImpl<TorchState> {
  final TorchStateLiveDataProxyApi api;

  TorchStateLiveDataImpl.internal(this.api);

  @override
  Future<TorchState?> getValue() => api.getValue().then((e) => e?.impl);

  @override
  Future<void> observeForever(Observer<TorchState> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<TorchState> observer) =>
      api.removeObserver(observer.api);
}

final class ZoomStateLiveDataImpl extends LiveDataImpl<ZoomState> {
  final ZoomStateLiveDataProxyApi api;

  ZoomStateLiveDataImpl.internal(this.api);

  @override
  Future<ZoomState?> getValue() => api.getValue().then((e) => e?.impl);

  @override
  Future<void> observeForever(Observer<ZoomState> observer) =>
      api.observeForever(observer.api);

  @override
  Future<void> removeObserver(Observer<ZoomState> observer) =>
      api.removeObserver(observer.api);
}

extension CameraStateLiveDataProxyApiX on CameraStateLiveDataProxyApi {
  LiveData<CameraState> get impl => CameraStateLiveDataImpl.internal(this);
}

extension TorchStateLiveDataProxyApiX on TorchStateLiveDataProxyApi {
  LiveData<TorchState> get impl => TorchStateLiveDataImpl.internal(this);
}

extension ZoomStateLiveDataProxyApiX on ZoomStateLiveDataProxyApi {
  LiveData<ZoomState> get impl => ZoomStateLiveDataImpl.internal(this);
}
